torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,134 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// TODO(osalpekar): Tune RetryDuration to be something in seconds per
+// team recommendations. Kept at 100ms for the sake of Milestone 1.
+const auto kRpcRetryDuration = std::chrono::milliseconds(100);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+const float kBackoff = 1.2;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(true);
+  }
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  {
+    std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+    rpcAgentRunning_.store(false);
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+}
 
-RpcAgent::~RpcAgent() = default;
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+  auto future = std::make_shared<FutureMessage>();
+  auto fm = send(to, std::move(message));
```

Comment:
you're holding rpcRetryMutex_ too broadly here, e.g. I don't think you should hold across send()